lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/SchemaResolver.java

```diff
@@ -0,0 +1,224 @@
+    schema.addProp(UR_SCHEMA_ATTR, name);
+    return schema;
+  }
+
+  static boolean isUnresolvedSchema(final Schema schema) {
+    return (schema.getType() == Schema.Type.RECORD && schema.getProp(UR_SCHEMA_ATTR) != null);
+  }
+
+  static String getUnresolvedSchemaName(final Schema schema) {
+    String name = schema.getProp(UR_SCHEMA_ATTR);
+    if (name == null) {
+      throw new IllegalArgumentException("Schema " + schema + " is not a unresolved schema");
+    } else {
+      return name;
+    }
+  }
+
+  static Protocol resolve(final Protocol protocol) {
+    Protocol result = new Protocol(protocol.getName(), protocol.getDoc(), protocol.getNamespace());
+    final Collection<Schema> types = protocol.getTypes();
+    List<Schema> newSchemas = new ArrayList(types.size());
+    Map<String, Schema> processed = new HashMap<String, Schema>();
+    for (Schema schema : types) {
+      newSchemas.add(resolve(schema, protocol, processed));
+    }
+    result.setTypes(newSchemas);
+
+    for (Map.Entry<String, Protocol.Message> entry : protocol.getMessages().entrySet()) {
+      Protocol.Message value = entry.getValue();
+      Protocol.Message nvalue;
+      if (value.isOneWay()) {
+        Schema request = value.getRequest();
+        nvalue = result.createMessage(value.getName(), value.getDoc(),
+                value.getObjectProps(), intern(request, processed));
+      } else {
+        Schema request = value.getRequest();
+        Schema response = value.getResponse();
+        Schema errors = value.getErrors();
+        nvalue = result.createMessage(value.getName(), value.getDoc(),
+                value.getObjectProps(), intern(request, processed),
+                intern(response, processed), intern(errors, processed));
+      }
+      result.getMessages().put(entry.getKey(), nvalue);
+    }
+    copyProps(protocol, result);
+    return result;
+  }
+
+  private static void copyProps(final JsonProperties from, final JsonProperties to) {
+    for (Map.Entry<String, JsonNode> entry : from.getJsonProps().entrySet()) {
+      to.addProp(entry.getKey(), entry.getValue());
+    }
+  }
+
+  static Schema resolve(final Schema schema, final Protocol protocol, final Map<String, Schema> processed) {
+    final String fullName = schema.getFullName();
+    if (fullName != null && processed.containsKey(fullName)) {
+      return processed.get(schema.getFullName());
+    } else if (isUnresolvedSchema(schema)) {
+      final String unresolvedSchemaName = getUnresolvedSchemaName(schema);
+      Schema type = protocol.getType(unresolvedSchemaName);
+      if (type == null) {
+        throw new IllegalArgumentException("Cannot resolve " + unresolvedSchemaName);
+      }
+      return resolve(type, protocol, processed);
+    } else {
+      switch (schema.getType()) {
+        case RECORD:
+          Schema createRecord = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(),
```

Comment:
I prefer to make new instances(copies) instead of mutating an existing object. (mostly out of principle)
A schema visitor utility would be useful in any case...